File: 2140_solving_questions_with_brainpower.swift
// https://leetcode.com/problems/solving-questions-with-brainpower/

class Solution {
    var memo = [Int]()

    func mostPoints(_ questions: [[Int]]) -> Int {
        memo = Array(repeating: 0, count: questions.count)

        for i in stride(from: questions.count - 1, through: 0, by: -1) {
            let skipScore = getRestPoints(index: i + 1)
            let solveScore = questions[i][0] + getRestPoints(index: i + 1 + questions[i][1])
            memo[i] = max(skipScore, solveScore)
        }

        return memo[0]
    }

    private func getRestPoints(index: Int) -> Int {
        if index >= memo.count {
            return 0
        } else {
            return memo[index]
        }
    }
}

File: 2140_solving_questions_with_brainpower.kt
// https://leetcode.com/problems/solving-questions-with-brainpower/description/

import java.lang.Long.max

class Solution {
    fun mostPoints(questions: Array<IntArray>): Long {
        val n = questions.size
        val dp = LongArray(n)
        for (i in n - 1 downTo 0) {
            val skipPoint = getPoint(i + 1, dp)
            val doPoint = getPoint(i + 1 + questions[i][1], dp) + questions[i][0].toLong()
            dp[i] = max(skipPoint, doPoint)
        }
        return dp[0]
    }

    private fun getPoint(i: Int, dp: LongArray): Long {
        if (i >= dp.size) {
            return 0
        } else {
            return dp[i]
        }
    }
}
